src/test/app/Batch_test.cpp

```diff
@@ -0,0 +1,3047 @@
+            env.close();
+            validateBatch(env, parentBatchId, testCases);
+
+            BEAST_EXPECT(env.balance(alice) == preAlice - XRP(200) - batchFee);
+            BEAST_EXPECT(env.balance(bob) == preBob + XRP(100));
+            BEAST_EXPECT(env.balance(carol) == preCarol + XRP(100));
+        }
+
+        {  // Only one: the fourth succeeds
+            auto const preAlice = env.balance(alice);
+            auto const preBob = env.balance(bob);
+            auto const preCarol = env.balance(carol);
+            auto const seq = env.seq(alice);
+            auto const batchFee = batch::calcBatchFee(env, 0, 6);
+            env(batch::outer(alice, seq, batchFee, tfOnlyOne),
+                batch::inner(
+                    offer(
+                        alice,
+                        alice["USD"](100),
+                        XRP(100),
+                        tfImmediateOrCancel),
+                    seq + 1),
+                batch::inner(
+                    offer(
+                        alice,
+                        alice["USD"](100),
+                        XRP(100),
+                        tfImmediateOrCancel),
+                    seq + 2),
+                batch::inner(
+                    offer(
+                        alice,
+                        alice["USD"](100),
+                        XRP(100),
+                        tfImmediateOrCancel),
+                    seq + 3),
+                batch::inner(pay(alice, bob, XRP(100)), seq + 4),
+                batch::inner(pay(alice, carol, XRP(100)), seq + 5),
+                batch::inner(pay(alice, eve, XRP(100)), seq + 6));
+
+            auto const txIDs = env.tx()->getBatchTransactionIDs();
+            TxID const parentBatchId = env.tx()->getTransactionID();
+            std::vector<TestBatchData> testCases = {
+                {"tecKILLED", to_string(txIDs[0])},
+                {"tecKILLED", to_string(txIDs[1])},
+                {"tecKILLED", to_string(txIDs[2])},
+                {"tesSUCCESS", to_string(txIDs[3])},
+            };
+            env.close();
+            validateBatch(env, parentBatchId, testCases);
+
+            BEAST_EXPECT(env.balance(alice) == preAlice - XRP(100) - batchFee);
+            BEAST_EXPECT(env.balance(bob) == preBob + XRP(100));
+            BEAST_EXPECT(env.balance(carol) == preCarol);
+        }
+    }
+
+    void
+    testInnerSubmitRPC(FeatureBitset features)
+    {
+        testcase("inner submit rpc");
+
+        using namespace test::jtx;
+        using namespace std::literals;
+
+        test::jtx::Env env{*this, envconfig()};
+
+        auto const alice = Account("alice");
+        auto const bob = Account("bob");
+
+        env.fund(XRP(1000), alice, bob);
+        env.close();
+        // Invalid RPC Submission: TxnSignature
+        // - has `TxnSignature` field
+        // - has no `SigningPubKey` field
+        // - has no `Signers` field
+        // - has `tfInnerBatchTxn` flag
+        {
```

Comment:
I think you could simplify this section a fair amount with a lambda that does the submitting/checking instead of needing to rewrite it every time.